### ISSUE TITLE
[Issue] CVE-2023-23397.ps1 fails to process DiscoverySearchMailbox

### ISSUE THREAD

**Issue by lusassl-msft** [microsoft/CSS-Exchange#1564]
**Describe the issue**
Reported by a user via blog post comment:
```

cmdlet CVE-2023-23397.ps1 at command pipeline position 2
Supply values for the following parameters:
Credential
CVE-2023-23397 script version 23.03.14.1625
Exception setting "Credentials": "Cannot convert the "Microsoft.Exchange.WebServices.Data.WebCredentials" value of type "Microsoft.Exchange.WebServices.Data.WebCredentials" to type "Microsoft.Exchange.WebServices.Data.ExchangeCredentials"."
At C:\Users\Administrator.DOMAIN\Desktop\CVE-2023-23397.ps1:1050 char:13
+ $Service.Credentials = New-Object Microsoft.Exchange.WebS ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo : NotSpecified: (:) [], SetValueInvocationException
+ FullyQualifiedErrorId : ExceptionWhenSetting

 

Exception setting "ImpersonatedUserId": "Cannot convert the "Microsoft.Exchange.WebServices.Data.ImpersonatedUserId" value of type "Microsoft.Exchange.WebServices.Data.ImpersonatedUserId" to type "Microsoft.Exchange.WebServices.Data.ImpersonatedUserId"."
[03/14/2023 19:31:34] : Scanning 3 of 7 mailboxes (currently: DiscoverySearchMailbox{GUID2}@domain.com)
[03/14/2023 19:31:34] : Unable to process mailbox DiscoverySearchMailbox{DiscoverySearchMailbox{GUID2}@domain.com, Either you don't have proper impersonation rights or the mailbox is inaccessible. Inner Exception:


```

**Comment by MrReed** [comment]
Hi, for errors like "Either you don't have proper impersonation rights or the mailbox is inaccessible." check the steps under "Add impersonation rights in Exchange admin center (EAC)" on https://www.codetwo.com/kb/how-to-set-impersonation-rights-manually/

Don't forget to assign the modified role to a user e.g. the tenant admin.

**Comment by brepmasterdav** [comment]
OK, that makes sense. I used the examples [here ](https://microsoft.github.io/CSS-Exchange/Security/CVE-2023-23397/)under the Exchange Online section. Does that need amending?

Should it be `Get-RemoteMailbox | .\CVE-2023-23397.ps1 -Environment "Online"` instead?

**Comment by dpaulson45** [comment]
> OK, that makes sense. I used the examples [here ](https://microsoft.github.io/CSS-Exchange/Security/CVE-2023-23397/)under the Exchange Online section. Does that need amending?
> 
> Should it be `Get-RemoteMailbox | .\CVE-2023-23397.ps1 -Environment "Online"` instead?

No, you just need to be in the correct shell session. 

If you are running against EXO, you should be in an EXO shell session.

**Comment by dpaulson45** [comment]
> **Describe the issue** Reported by a user via blog post comment:
> 
> ```
> 
> cmdlet CVE-2023-23397.ps1 at command pipeline position 2
> Supply values for the following parameters:
> Credential
> CVE-2023-23397 script version 23.03.14.1625
> Exception setting "Credentials": "Cannot convert the "Microsoft.Exchange.WebServices.Data.WebCredentials" value of type "Microsoft.Exchange.WebServices.Data.WebCredentials" to type "Microsoft.Exchange.WebServices.Data.ExchangeCredentials"."
> At C:\Users\Administrator.DOMAIN\Desktop\CVE-2023-23397.ps1:1050 char:13
> + $Service.Credentials = New-Object Microsoft.Exchange.WebS ...
> + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> + CategoryInfo : NotSpecified: (:) [], SetValueInvocationException
> + FullyQualifiedErrorId : ExceptionWhenSetting
> 
>  
> 
> Exception setting "ImpersonatedUserId": "Cannot convert the "Microsoft.Exchange.WebServices.Data.ImpersonatedUserId" value of type "Microsoft.Exchange.WebServices.Data.ImpersonatedUserId" to type "Microsoft.Exchange.WebServices.Data.ImpersonatedUserId"."
> [03/14/2023 19:31:34] : Scanning 3 of 7 mailboxes (currently: DiscoverySearchMailbox{GUID2}@domain.com)
> [03/14/2023 19:31:34] : Unable to process mailbox DiscoverySearchMailbox{DiscoverySearchMailbox{GUID2}@domain.com, Either you don't have proper impersonation rights or the mailbox is inaccessible. Inner Exception:
> ```

This issue is resolved by the following:

- Download correct version: [https://api.nuget.org/v3-flatcontainer/microsoft.exchange.webservices/2.2.0/microsoft.exchange.webservices.2.2.0.nupkg](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapi.nuget.org%2Fv3-flatcontainer%2Fmicrosoft.exchange.webservices%2F2.2.0%2Fmicrosoft.exchange.webservices.2.2.0.nupkg&data=05%7C01%7Cdpaul%40microsoft.com%7C289e8d0ea6c64613cb9808db259e5a22%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638145134343202177%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=N%2BVlGk0XpY04%2BMXIsE3etfNzvmqz30VzFi1HRJxda%2B4%3D&reserved=0)
- CLOSE POWERSHELL (don’t know why but this is a must otherwise you continue to get the same error over and over again) 
- Open new PowerShell session 
- Run the cmdlet point to the new DLL path for “lib\40\Microsoft.Exchange.WebServices.dll”

**Comment by lusassl-msft** [comment]
Resolved with PR #1575 as this changed the download Uri in documentation. The script downloads the right version of the machine on which the script runs has internet access.

**Comment by lusassl-msft** [comment]
Resolved. Please follow the steps outlined here to fix: https://microsoft.github.io/CSS-Exchange/Security/CVE-2023-23397/#cannot-convert-the-microsoftexchangewebservicesdatawebcredentials
